libyul/optimiser/VarNameCleaner.cpp

```diff
@@ -0,0 +1,111 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <libyul/optimiser/VarNameCleaner.h>
+#include <libyul/AsmData.h>
+#include <algorithm>
+#include <cctype>
+#include <climits>
+#include <iterator>
+#include <string>
+
+using namespace yul;
+using namespace std;
+
+// ensure the following transforms
+// - input: a, a_1, a_1_2
+//	 output: a, a_1, a_2
+// - input: a, a_1, a_1_2, a_2
+//   output: a, a_1, a_2, a_3
+// - input: a_15, a_17
+//   output: a, a_2
+// - input: abi_decode_256
+//   output: abi_decode
```

Comment:
This shows this may be a legitimate name.

I think we may want to consider using a separator before SSA indexes, such as `a$1`, `a$2`, etc.

Note, #4108 proposes to support `.` for identifiers, such as `evm.return` and the `$` wouldn't conflict with that.

Reply:
If we really want to retain names, then I think the proper way is to escape them as the very first step, unless we want to introduce a character that is only valid during optimization (note that in this case, we cannot export to string and re-import, which is bad, I would say).

In any case, I think this should be deferred to another issue.